src/index.js

```diff
@@ -135,13 +135,22 @@ function matchesSelector (element, ast) {
       if (node.value === ' ') {
         // walk all ancestors
         const precedingNodes = getLastNonCombinatorNodes(nodes.slice(0, i))
-        const ancestor = getFirstMatchingAncestor(element, precedingNodes)
+        const prePrecedingNodes = nodes.slice(0, i - precedingNodes.length)
+        let ancestor = getFirstMatchingAncestor(element, precedingNodes)
         if (!ancestor) {
           return false
-        } else {
-          element = ancestor
-          i -= precedingNodes.length
         }
+        // Even though first ancestor matches the selector, ancestor should match all preceding nodes
```

Comment:
After thinking about it a bit more, I think I understand why this code works, but it would be really helpful to have a long comment here explaining why we need a loop-within-a-loop. Something like:

```suggestion
        // Even though the first ancestor matches the selector, the ancestor should match all preceding nodes.
        // For example, consider a tree like `body > div > div > div > button`.
        // The selector `body > div button` should match the button, but it won't if we simply grab the first
        // div that is an ancestor of the button. Instead, we have to keep walking up the tree, trying all divs, until
        // we find one that matches the *previous selectors*, i.e. until we match the div that is the immediate
        // child of the body.
        // Note that this can get quite expensive, as it is an O(n^2) algorithm, but it is correct. Also, this algorithm
        // only needs to be implemented for the space (`' '`) combinator - the `'>'` combinator only needs to
        // look at the immediate parent, so it doesn't need to iterate through all possible matching ancestors.
```

Reply:
I moved this part of code to a new function to accommodate these long comments. Hope this is fine?